### PR TITLE
Have content loader manage framework messages

### PR DESCRIPTION
**Pull request by quis**
This is to let us manage pieces of content which are specific to a combination
of:
- a place on a page
- a framework
- the status of that framework
- the status of a supplier‘s application to a framework

eg, on the supplier dashboard there is a block telling you about your,
declaration which has different content through different stages of the process.
Currently we manage this with lots of {% if %} statements in the templates,
which is:
- hacky
- not scaleable

These pieces of content will be stored in the frameworks repo, in a directory
called 'messages', eg
`./frameworks/g-cloud-7/messages/index.yml`

These file will contain a dictionary of statuses, eg
```
g-cloud-7/messages/index.yml
---
- open-not_interested:
    message: please register interest
- open-registered_interest:
    message: please complete your application
- pending-registered_interest:
    message: you only did the registering interest bit
```

The format of the filename is:
`<block>.yml`

### CODE REVIEW

File: dmutils/content_loader.py [hunk 1]
```diff
@@ -369,6 +369,8 @@ class ContentLoader(object):
     >>> loader.load_manifest('framework-2', 'manifest-1', 'question-set-1')
     >>> # get a builder
     >>> loader.get_builder(framework_slug, 'manifest-1')
+    >>> # get a message
+    >>> messages.get(framework_slug, 'homepage_sidebar', 'in_review')
```
Comment (robyoung) [inline dmutils/content_loader.py]: I think this should be
```python
>>> loader.get_message(framework_slug, 'homepage_sidebar', 'in_review')
```

File: dmutils/content_loader.py [hunk 2]
```diff
@@ -416,6 +418,27 @@ def get_question(self, framework_slug, question_set, question):
 
         return self._questions[framework_slug][question_set][question].copy()
 
+    def get_message(
+        self, framework_slug, block, framework_status, supplier_status=None
+    ):
+
+        framework = self._content[framework_slug]
+
+        if 'messages' not in framework:
```
Comment (allait) [inline dmutils/content_loader.py]: I feel like I'm missing some context here.
As far as I understand `_content[framework_slug]` is a dictionary of manifest files (which are lists of sections), why are we adding a special non-manifest dictionary item to it? And what does the message yaml file look like? I have a suspicion that `ContentBuilder` won't really apply to them.

File: dmutils/content_loader.py [hunk 3]
```diff
@@ -432,6 +458,20 @@ def _populate_section(self, framework_slug, question_set, section):
...
+        self, block, framework_status, supplier_status
+    ):
+        return '{}-{}{}'.format(
+            block,
+            framework_status,
+            '-{}'.format(supplier_status) if supplier_status else ''
+        )
+
+    def _message_error(self, framework_slug, message):
+        self._content[framework_slug]['messages'][message] = None
```
Comment (allait) [inline dmutils/content_loader.py]: Why is this needed? No one should be catching the `ContentNotFoundError`, so the process will termniate once the exception is thrown.

**Comment by allait** [comment]
What does the content of each `<block>-<status>-<optional:supplier_status>.yml` file look like? Is there a way we could join them into `<block>.yml` and avoid having an additional structure of nested keys within the file names?

**Comment by pcraig3** [comment]
Here's what I've done so far as far as messages go: https://github.com/alphagov/digitalmarketplace-frameworks/commit/4d6fffec066d6fdb05f4360f05e9468ca9028df7

File: dmutils/content_loader.py (hunk 1 above)
Comment (quis) [inline dmutils/content_loader.py]: Fixed this and squashed.

**Comment by allait** [comment]
@pcraig3 they look small enough so far to be merged into a single `index.yml` with a top-level framework status dictionary and avoid the name formatting

File: dmutils/content_loader.py (hunk 3 above)
Comment (quis) [inline dmutils/content_loader.py]: Intention is that if there isn't a file for a specific block and status then the front end won't render anything. So the app needs to be able to ask for a file which doesn't exist without throwing an exception. And we should cache this as `None` so that it doesn't make a disk operation every time it does this.

File: dmutils/content_loader.py (hunk 2 above)
Comment (quis) [inline dmutils/content_loader.py]: `ContentBuilder` won't apply. Maybe the messages should be stored in their own `dict`.

File: dmutils/content_loader.py (hunk 3 above)
Comment (allait) [inline dmutils/content_loader.py]: This sounds like another reason to merge the files together. I think that "all files that we attempt to load should exist" is a nice property and it would be good if we could keep it.

File: dmutils/content_loader.py (hunk 2 above)
Comment (allait) [inline dmutils/content_loader.py]: Yeah, `.get_builder` would be the only reason to add things to `_content`. We can just add a new property to store message.

**Comment by quis** [comment]
@allait rewrote things to:
- store messages separately from questions
- remove the statuses from filenames, leaving them as `<block>.yml`

**Comment by allait** [comment]
We could split `get_message` from `load_message` so that we can read the message files on startup similar to how we deal with manifests. It's possible to just call `.get_message` for this, but it would be a bit less clear.

**Comment by quis** [comment]
How would we read in all the messages? Walk the directory? Have some kind of manifest? As a call to the constructor?

Don't see what's too bad about reading them as and if needed.

**Comment by allait** [comment]
@quis we shouldn't have that many blocks so listing them the same way we do manifests should work. There's nothing a running app can do about a missing file, so it makes sense to make sure everything is in place before starting. Reading files when needed isn't a big problem as long as all related views have CI tests though.

**Comment by quis** [comment]
Would it be better if content_loader was just given a list of frameworks, eg
``` python
frameworks = api.get_all_the_frameworks()  # ['g-cloud-7', 'g-cloud-8', 'dos-2', …]
content_loader = ContentLoader(frameworks)
```

Then we read each manifest for that framework, plus any messages that exist?

**Comment by allait** [comment]
@quis the reason we list manifests (and schema files in the API) is that it forces the app to load the given filename, while e.g. just listing the directory doesn't mean that the file we actually expect to exist is there.

**Comment by quis** [comment]
Suppose it's OK while we have a small number of frameworks. Added `.load_messages()` in 659a1ac.

**Comment by allait** [comment]
:+1:
